### PR TITLE
Improve the inserter a11y

### LINKED ISSUE #525: Inserter a11y improvements
See #515

The inserter toggle button and menu should use some ARIA roles, states, and properties to improve both the feedback for users and the actual interaction. For example:

- the toggle button doesn't inform users that it's meant to open a menu, should use `aria-haspopup`
- when activating the toggle button, there's no audible feedback about what happened, should use `aria-expanded`
- the menu and menu items should use `role="menu"` and `role="menuitem"`, this is not just for better semantics but also because we need to stop screen readers from intercepting keypresses, otherwise arrow navigation won't work (see #515)
- the search field needs an explicitly associated label with for/id attributes (can be visually hidden)
- the menu items need a focus style

I've played a bit with this, will do a PR soon.

### PR DESCRIPTION
**Pull request by afercia**
Improves the inserter accessibility.

- renames 'Common' to 'Common blocks', to give better context (by the way, I guess this should be made translatable?)
- adds `aria-haspopup="true"` and `aria-expanded` to the inserter toggle button
- adds `role="menu"` and `aria-labelledby` to the inserter menu, and necessary IDs
- adds `tabIndex="0"` to the menu to make it focusable, this way it gets announced as "menu" with the name provided by `aria-labelledby`; this will work with multiple menus too, each one will be announced with the `category.title`
- adds `role="menuitem"` to the inserter menu items
- adds a properly associated `label` to the search field, adding necessary ID
- fixes toggle button position in Safari 10
- adjusts menu max-height (I guess this would need further adjustments)
- improves menu items text alignment in Safari 10
- adds focus style on the menu items, same as hover

*Important note*:
when #515 will be done, the menu items will need `tabIndex="-1"`, as focus should be managed programmatically for the Tab and Arrows keys

Fixes #525

### CODE REVIEW

File: editor/components/inserter/menu.js
Comment (aduth) [inline editor/components/inserter/menu.js]: Can you clarify whether the following would work just as well to associate the label with the input?

```jsx
<label>
	<span className="screen-reader-text">
		{ wp.i18n.__( 'Search blocks' ) }
	</span>
	<input />
</label>
```

I only raise in an effort to avoid `id`, since it's hard to guarantee that a component will only occur once in a page.

Same applies to the above `aria-labelledby`, but an alternative is less clear to me. In cases where I've required an ID in a React component, it can work to store a static incrementing number that's assigned in mount:

```jsx
class Foo extends Component {
	constructor() {
		super( ...arguments );
		this.instanceId = this.constructor.instances++;
	}

	render() {
		return <span id={ 'foo-' + this.instanceId } />;
	}
}

Foo.instances = 0;
```

But, evidenced by this example, it's a bit cumbersome to manage.

File: editor/components/inserter/style.scss
Comment (aduth) [inline editor/components/inserter/style.scss]: - We can drop the `&` and the nesting would work just as well.
- I had a hard time finding what `.insert` was in this case; found later that it's the Dashicon. For future readers, maybe being overly specific could clarify, e.g. `.dashicon.insert`

File: editor/components/inserter/style.scss
Comment (aduth) [inline editor/components/inserter/style.scss]: - Would we want the comment to appear in the generated stylesheet? With `/*` it will, otherwise SASS's `//` could prevent it from being included
- What specific to Safari 10 is this addressing? Might help future maintainers decide whether this can be dropped.

**Comment by afercia** [comment]
Giving the inserter menu and items proper ARIA roles, will allow keyboard events to correctly work when using screen readers, see #515.

Making the menu focusable and using `aria-labelledby` will make screen readers announce it as "menu" and with the name used for the visible menu "title". To avoid the title gets announced twice, the visible one has a `aria-hidden="true"` attribute.  `aria-labelledby` works even if it targets and element with `aria-hidden="true"`. Screenshots:

Safari 10 + VoiceOver:

menu aria-labelledby items count

How `aria-labelledby` works: see in the Safari inspector:

aria-labelledby in practice

or in IE 11 and AViewer:

ie11 menu aria-labelledby

Notice above the SVG titles are announced as part of the menu items in IE11, will open a separate issue specific to SVG accessibility.

The `menu` and `menuitem` roles give also one more bonus: some browsers and screen readers get able to announce the number of items, see Firefox + NVDA:

inserter firefox nvda

File: blocks/api/categories.js
Comment (aduth) [inline blocks/api/categories.js]: Not necessarily in scope of this pull request, but the string ought to be translated.

Comment (afercia) [inline blocks/api/categories.js]: > by the way, I guess this should be made translatable? 🙂

File: editor/components/inserter/menu.js
Comment (afercia) [inline editor/components/inserter/menu.js]: There are two ways to associate labels to form controls: implicit and explicit. Implicit: the label wraps the control and doesn't use a `for` attribute. Explicit: label and control are separate elements (no wrapping) and they use for/id. The latter is the recommended one for accessibility. Never do a mix of the two techniques 🙂 

> Same applies to the above aria-labelledby

ARIA attributes often need IDs. We'll need lots of IDs 🙂  Happy to implement any best practice, but as you pointed out in some cases it will be a bit hard.

Comment (aduth) [inline editor/components/inserter/menu.js]: >The latter is the recommended one for accessibility. Never do a mix of the two techniques 🙂

For my own curiosity, could you clarify the "why" of both of these two points?

**Comment by afercia** [comment]
To clarify the `.dashicon.insert` change, seems Safari 10 doesn't like flex whit SVG. Without `margin: 0 auto` the icon is not centered:

screen shot 2017-04-27 at 19 00 23

About the other buttons (the ones with a SVG icon and text), the button uses `display: flex` but one of the flex-items is actually a text node:

screen shot 2017-04-27 at 19 02 54

Seems in some browsers this works, but Safari didn't align vertically the text, I had to add some `line-height`. As far as I know, flex applies to elements, not to text nodes. Not sure this is actually an issue, but maybe something worth considering.

File: editor/components/inserter/menu.js
Comment (afercia) [inline editor/components/inserter/menu.js]: In short, explicit labels are the ones better supported by browsers/screen readers combos. See also: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/#labeling

A mix of the two techniques is against the spec and in some combos could make the label being announced twice.

**Comment by afercia** [comment]
IDs are now generated. See in the screenshot below, with the top and bottom inserters open at the same time (actually, there should be only one instantiated at a time, inserters should close when clicking anywhere outside of them, see https://github.com/WordPress/gutenberg/issues/515#issuecomment-297701709):

screen shot 2017-04-27 at 22 02 31

**Review by aduth — APPROVED** [review]
This looks great 👍  It seems the translation files have drifted from master a bit, so I'm going to rebase and force push an update shortly before merging.

**Comment by afercia** [comment]
Yep I was in doubt what to do with the gutenberg.pot, wouldn't be better to ecxlude it from commits?

**Comment by aduth** [comment]
>Yep I was in doubt what to do with the gutenberg.pot, wouldn't be better to ecxlude it from commits?

We probably need a better way to automate this. I'd considered a bot generating and committing in response to a master commits webhook, but there's not a great long-term core reconciliation story to this approach.

**Comment by aduth** [comment]
In response to noticing that it wasn't just the Inserter button that was misaligned but also Undo / Redo, I started down a rabbit hole of adjusting the icon centering approach by using padding on the underlying `IconButton`. However this had unintended consequences on the toolbar controls, specifically those with adjacent subscript text (heading levels, quote styles). I'd argue it more surfaced some preexisting issues. One being that there is icon-specific whitespace for some icons and not others (like the heading H icon).

The latest commit works somewhat well; could probably do for a design review @jasmussen . If it's too much of a detour, we could come up with a simpler alternative as a short-term solution, especially since it's not particularly in scope to the intent of the pull request.

**Comment by aduth** [comment]
Related to `.pot` discussion: #604

**Comment by aduth** [comment]
Rebased and plan to merge this as-is. Design concerns around icon button centering can be addressed separately, though for what it's worth, the same issue exists already in master with quote style subscript proximity to the icon itself.

`master`|`update/better-inserter-a11y`
---|---
master|update/better-inserter-a11y

**Pull request merged**
